### PR TITLE
build: convert from JavaScript to TypeScript

### LINKED ISSUE #6: Use TypeScript instead of JavaScript for code/tests
**What would you like node-schedule to be able to do?**

It's probably a good idea to use TypeScript instead of JavaScript, so that we're consistent with our programming language between this repository and the main Renovate repository.

As an added bonus we'll probably have a safer programming language to work in, so that we can catch stupid stuff earlier. 😄 

**Did you already have any implementation ideas?**

I've taken a look at the migration guide from JavaScript to TypeScript in the official Typescript docs, it shouldn't be that hard for a programmer familiar with TypeScript to migrate this repository and tooling.

I'm not that good with TypeScript (or JavaScript for that matter 🙈 ) to be able to migrate us. So I'd like somebody else to do this, or maybe we could do it together, so that I can learn a bit (but that'll slow you down...).

[TypeScript docs on migrating from JavaScript](https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html)

### PR DESCRIPTION
**Pull request by patlehmann1**
## Changes:

This is the conversion of the `index.js` and `test.js` files to typescript. Set up the typescript compiler and `ts-jest` so that typescript tests can be used.

## Context:

Closes #6 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -46,6 +46,9 @@ build/Release
 node_modules/
 jspm_packages/
 
+# typescript build from compiling
+built/
```
Comment (HonkingGoose) [inline .gitignore:50]: Are you sure the correct name for this directory is `built`?
I think usually the output directory is called `build`, so with a `d` at the end?

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    "strictNullChecks": true,
+    "target": "es6"
+  },
+  "include": ["./src/**/*"],
+  "exclude": [
+    "./node_modules/**/*",
+    "./node_modules/**/**",
+    "./jest.config.js",
+  ]
+}
```
Comment (HonkingGoose) [inline tsconfig.json:23]: Seems we have a end of line (EOL) error in this file?

**Review by HonkingGoose — COMMENTED** [review]
Thank you for your work! 🚀 ❤️ 

I've requested a review from one of the Renovate maintainers, as they know more about TypeScript than me.

From looking at it, it seems mostly OK, but I haven't yet ran the code. 😄 

I do have some things that I'm not 100% sure about.

File: yarn.lock [hunk 3]
```diff
@@ -4,19 +4,19 @@
 
 "@babel/code-frame@^7.12.13", "@babel/code-frame@^7.14.5":
   version "7.14.5"
-  resolved "https://registry.yarnpkg.com/@babel/code-frame/-/code-frame-7.14.5.tgz#23b08d740e83f49c5e59945fbf1b43e80bbf4edb"
+  resolved "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.14.5.tgz"
```
Comment (HonkingGoose) [inline yarn.lock:7]: I don't know why the registry is changing here?

File: .gitignore (hunk 1 above)
Comment (patlehmann1) [reply .gitignore:50]: I just use conventions from work that I do at my job. and we call the folder `built`. I can make the change though, no problem.

File: tsconfig.json (hunk 2 above)
Comment (patlehmann1) [reply tsconfig.json:23]: I will take a look, I didn't get any errors when working on it.

File: package.json [hunk 4]
```diff
@@ -3,7 +3,7 @@
   "version": "0.0.1",
   "private": "true",
   "description": "Node schedule utilities for use with Renovate bot.",
-  "main": "index.js",
+  "main": "built/src/index.js",
```
Comment (HonkingGoose) [inline package.json:6]:
```suggestion
  "main": "build/src/index.js",
```

I think this needs to change as well?

File: package.json [hunk 5]
```diff
@@ -28,10 +28,13 @@
   "keywords": [],
   "license": "MIT",
   "devDependencies": {
+    "@types/jest": "^26.0.23",
```
Comment (HonkingGoose) [inline package.json:31]:
```suggestion
    "@types/jest": "26.0.23",
```

We pin all the versions we use in our `package.json`. 😉

File: package.json [hunk 6]
```diff
@@ -28,10 +28,13 @@
...
     "cross-env": "7.0.3",
     "jest": "27.0.6",
     "jest-silent-reporter": "0.5.0",
-    "prettier": "2.3.2"
+    "prettier": "2.3.2",
+    "ts-jest": "^27.0.3",
```
Comment (HonkingGoose) [inline package.json:36]:
```suggestion
    "ts-jest": "27.0.3",
```

File: package.json [hunk 7]
```diff
@@ -28,10 +28,13 @@
...
+    "typescript": "^4.3.5"
```
Comment (HonkingGoose) [inline package.json:37]:
```suggestion
    "typescript": "4.3.5"
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some more things, sorry about that! 😉

File: package.json (hunk 4 above)
Comment (patlehmann1) [reply package.json:6]: Yes, I missed that one. Good catch.

**Comment by HonkingGoose** [comment]
@patlehmann1 You'll need to run `yarn` again to create the lockfile with the correct pinned dependencies. 😉

File: jest.config.js [hunk 8]
```diff
@@ -9,4 +9,26 @@ module.exports = {
   collectCoverage: true,
   coverageDirectory: "./coverage",
   coverageReporters: ["html", "json", "text-summary"],
+  moduleFileExtensions: [
+    'js',
+    'json',
+    'node',
+    'ts'
+  ],
```
Comment (viceice) [inline jest.config.js:17]:
```suggestion
```

Included in `ts-jest` preset

File: jest.config.js [hunk 9]
```diff
@@ -9,4 +9,26 @@ module.exports = {
...
+      }
+    },
+    preset: 'ts-jest/presets/js-with-ts',
+    testMatch: [
+      '**/__tests__/*.test.[jt]s'
+    ],
+    testPathIgnorePatterns: [
+      '/build/'
+    ]
+  }]
```
Comment (viceice) [inline jest.config.js:33]:
```suggestion
    preset: 'ts-jest/presets/js-with-ts',    
    modulePathIgnorePatterns: ['<rootDir>/dist/', '/__fixtures__/'],
```

simplify

File: package.json [hunk 10]
```diff
@@ -3,7 +3,7 @@
   "version": "0.0.1",
   "private": "true",
   "description": "Node schedule utilities for use with Renovate bot.",
-  "main": "index.js",
+  "main": "build/src/index.js",
```
Comment (viceice) [inline package.json:6]:
```suggestion
  "main": "build/index.js",
```

re need to remove the additional `src` path

**Review by viceice — CHANGES_REQUESTED** [review]
I think i will do the required changes myself, as we like to have this repo configured like our main renovate repo.

**Comment by HonkingGoose** [comment]
@viceice do you want us to stop committing to this branch then, so that you can clean up?

**Comment by viceice** [comment]
you should at least push the required lockfile changes. i hope i can fix the other things later today

**Comment by HonkingGoose** [comment]
> you should at least push the required lockfile changes. i hope i can fix the other things later today

Hi @viceice! 👋 

I fixed the yarn lock file and ran Prettier. You can add your own fixes now. 😉

**Comment by HonkingGoose** [comment]
@viceice Are we ready for a final review? Or do you want/need to do more work?

**Review by viceice — APPROVED** [review]
the `tsconfig.lint.json` is for adding eslint in a later pr

File: .gitignore [hunk 11]
```diff
@@ -120,3 +120,9 @@ dist
 .yarn/build-state.yml
 .yarn/install-state.gz
 .pnp.*
+
```
Comment (HonkingGoose) [inline .gitignore:123]:
```suggestion
```

We could remove a newline here, that way it's consistent with the rest.

File: e2e/cjs/index.js [hunk 12]
```diff
@@ -0,0 +1,7 @@
+const ns = require('node-schedule');
+
+if (!ns.isEOL('v0.10.48')) {
+  throw new Error('Expected v10 to be EOL');
+}
+
+console.log('CommonJS e2e successful');
```
Comment (HonkingGoose) [inline e2e/cjs/index.js:7]:
```suggestion
const nodeSchedule = require('node-schedule');

if (!nodeSchedule.isEOL('v0.10.48')) {
  throw new Error('Expected v10 to be EOL');
}

console.log('CommonJS e2e successful');
```

I find a style like this to be more readable, what do you think?

File: e2e/cjs/package.json [hunk 13]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "schedule-test-cjs",
```
Comment (HonkingGoose) [inline e2e/cjs/package.json:2]: Seems we have a split between "normal JavaScript package" and "ESM style package"? I don't really understand this yet.

File: package.json [hunk 14]
```diff
@@ -1,45 +1,75 @@
...
+    "test-e2e:cjs": "npm pack && cd e2e/cjs && rimraf ./node_modules && npm install --no-package-lock --no-audit && yarn test",
+    "test-e2e:esm": "npm pack && cd e2e/esm && rimraf ./node_modules && npm install --no-package-lock --no-audit && yarn test"
   },
   "author": {
     "name": "HonkingGoose"
   },
   "contributors": [
-    "Lars-Magnus Skog <ralphtheninja@riseup.net>"
+    "Lars-Magnus Skog <ralphtheninja@riseup.net>",
+    "Michael Kriese <michael.kriese@visualon.de>"
```
Comment (HonkingGoose) [inline package.json:38]: Maybe @patlehmann1 wants to be listed as a contributor as well, as they did the initial migration work? 😉

File: tools/compile-schedule.js [hunk 15]
```diff
@@ -0,0 +1,5 @@
+import fs from 'fs/promises';
+
+const json = await fs.readFile('./schedule.json', { encoding: 'utf8' });
+
+await fs.writeFile('./src/schedule.js', `export default ${json}`);
```
Comment (HonkingGoose) [inline tools/compile-schedule.js:5]: Is this what makes the `schedule.json` from upstream node into something that Renovate bot can use in the main repo?

**Comment by viceice** [comment]
Ups, to late, 😅  you requested my review, so i thought you are done 🤷‍♂️

**Comment by HonkingGoose** [comment]
Yeah I was too late... 👻 

Do you care about any of my points or are you happy with what we have now? I can make a new PR with some changes. 😄

File: e2e/cjs/index.js (hunk 12 above)
Comment (viceice) [reply e2e/cjs/index.js:7]:
```suggestion
const { isEOL }= require('node-schedule');

if (!isEOL('v0.10.48')) {
  throw new Error('Expected v10 to be EOL');
}

console.log('CommonJS e2e successful');
```

What about this? It's more like the esm version.

File: e2e/cjs/package.json (hunk 13 above)
Comment (viceice) [reply e2e/cjs/package.json:2]: yes, the package is now compatible to be used in esm javascript module sytem, which is the new standard way or javascript for node js.

A lot packages are starting to be esm only, but renovate is not ready yet, so we publish in both types.

When renovate is ready for esm only, we can drop the commonjs version

File: package.json (hunk 14 above)
Comment (viceice) [reply package.json:38]: @patlehmann1 Feel free to open another pr to add yourself as contributor.

File: e2e/cjs/index.js (hunk 12 above)
Comment (HonkingGoose) [reply e2e/cjs/index.js:7]: That's also OK. I'll go make a PR for us to change this. 😉

File: tools/compile-schedule.js (hunk 15 above)
Comment (viceice) [reply tools/compile-schedule.js:5]: currently i use only the local `schedule.json` which can  be manuallly updated with `yarn get-latest-schedule-curl`.

In a future pr we can do some kind of auto updatting this file to trigger a ci release.

File: e2e/cjs/package.json (hunk 13 above)
Comment (HonkingGoose) [reply e2e/cjs/package.json:2]: Thank you for the explanation! 😄

**Pull request merged**